developer_docs/syntactic_stage.md

```diff
@@ -0,0 +1,82 @@
+
+The syntactic stage is described by the file [pyccel.parser.syntactic](../pyccel/parser/syntactic.py)
+
+The syntactic stage serves 4 main purposes:
+1.  Convert Python's [AST](https://docs.python.org/3/library/ast.html) (abstract syntax tree) representation of the python file to Pyccel's AST representation (objects of the classes in the folder [pyccel.ast](../pyccel/ast))
+2.  Raise an error for any syntax used that is not yet supported by pyccel
+3.  Convert header comments from strings to Pyccel's AST representation
+4.  Collect the name of all variables in each scope (see [scope](scope.md) for more details) to ensure no name collisions can occur if pyccel generates Variable names
+
+## Navigation and AST Creation
+
+The entry point for the class `SyntaxParser` is the function `parse`.
+This function is called from the constructor to examine a FST (Full Syntax Tree) object.
+The FST is the output of Python's `ast.parse` function with comments inserted using the [`pyccel.parser.extend_tree.extend_tree`](../pyccel/parser/extend_tree.py).
+Python's ast module discards all comments, so the function [`pyccel.parser.extend_tree.extend_tree`](../pyccel/parser/extend_tree.py) is needed to read the file and reinsert these comments at the correct location.
+
+The key line of the function `parse` is the call to `self._visit(self.fst)`.
+All elements of the tree must be visited.
+The `_visit` function internally calls a function named `_visit_X`, where `X` is the type of the object.
+This function must have the form:
+```python
+def _visit_ClassName(self, stmt):
+    ...
+```
+Each of these functions should internally call the `_visit` function on each of the elements of the object to obtain Pyccel AST nodes which can be combined to create a Pyccel AST node representing the current object.
+
+## Errors
+
+Error handling uses the classes in the file [pyccel.errors.errors](../pyccel/errors/errors.py).
+Errors in the syntactic stage should raise a `PyccelSyntaxError`.
+Where possible this should be done by accessing the `Errors()` singleton and calling the `report` function.
+This function takes several arguments (see docstring for more details).
+The most important arguments are:
+-   _message_ : Describe the issue that lead to the error
+
+-   _symbol_ : The Python ast object should be passed here. This object contains information about its position in the file (line number, column) which ensures the user can more easily locate their error
+
+-   _severity_ : The severity level must be one of the following:
+    -   _warning_ : An error will be printed but Pyccel will continue executing
+    -   _error_ : An error will be printed but Pyccel will continue executing the syntactic stage
+    -   _fatal_ : An error will be printed and Pyccel will stop executing. This level should rarely be needed in the syntactic stage as a failure in one function should not affect the execution of another. It is preferable to show the users all errors at once
```

Comment:
Is execution the right word here? it gives the impression that we are not in the syntactic stage, it has a _runtimy_ feel to it.

Reply:
I'm not sure what you mean? Are you perhaps confusing `pyccel` and `epyccel`? The executable `pyccel` stops executing. This is at runtime (although in python there is not really an alternative as there is no compile time)

Reply:
No, I meant that `execution of another` implies that a function is being executed, when in actuality it is rather syntacticly parsed or its syntactic parsing is executed not the function itself.  the following sounds better to me:
> This level should rarely be needed in the syntactic stage as a failure in __parsing__ one function should not affect the __parsing__ of another. It is preferable to show the users all errors at once

Reply:
Hmmm, now I see what you mean. But the functions I am talking about here are the `_visit_X` functions not the python functions being translated.  E.g. the failure of `_visit_Dict` shouldn't stop `_visit_List` working. You can have both a list and a dictionary in the same function

Reply:
How about:
> This level should rarely be needed in the syntactic stage as a failure in one visitation function should not affect the execution of another

Reply:
Sorry, I misunderstood. I thought you were taking a function as an example of what's beiing parsed. Now, I totally agree with you. What do you think of clearly stating that we are talking about `_visit_X` functions?

Reply:
I agree. That is what I tried to do by saying "visitation function". Is this not clear? Do you have a suggestion?

Reply:
> This level should rarely be needed in the syntactic stage as a failure in one visitation function (`_visit_X`) should not affect the execution of another

?

Reply:
Perfect!

Reply:
Fixed : d607b56b
(By the way, it is for you to decide when I have adequately responded to your comment and resolve the conversations once you're happy)